File: apps/mobile/lib/core/contracts/stores.dart
import 'dart:typed_data';

import 'trace_models.dart';
import 'trace_result.dart';

final class ObjectDraft {
  const ObjectDraft({
    required this.objectId,
    required this.tag,
    required this.reference,
  });

  final String objectId;
  final String tag;
  final ObjectReference reference;
}

abstract interface class ObjectStore {
  Future<TraceResult<ObjectReference>> create(ObjectDraft draft);
  Future<TraceResult<ObjectReference>> get(String objectId);
  Future<TraceResult<List<ObjectReference>>> getAllReferences();
  Future<TraceResult<void>> delete(String objectId);
}

abstract interface class SightingStore {
  Future<TraceResult<Sighting>> insert(Sighting sighting);
  Future<TraceResult<Sighting>> update(Sighting sighting);
  Future<TraceResult<Sighting?>> getLatest(String objectId);
  Future<TraceResult<List<Sighting>>> getTimeline(
    String objectId, {
    int limit = 50,
  });
}

enum SecureAssetType { referenceImage, sightingEvidence }

final class SecureAsset {
  const SecureAsset({
    required this.assetId,
    required this.type,
    required this.mimeType,
    required this.createdAt,
  });

  final String assetId;
  final SecureAssetType type;
  final String mimeType;
  final DateTime createdAt;
}

abstract interface class SecureAssetStore {
  Future<TraceResult<SecureAsset>> write({
    required SecureAssetType type,
    required Uint8List plaintext,
    required String mimeType,
  });

  Future<TraceResult<Uint8List>> read(String assetId);
  Future<TraceResult<void>> delete(String assetId);
}

File: apps/android/core/contracts/src/main/kotlin/com/traceapp/core/contracts/Stores.kt
package com.traceapp.core.contracts

data class ObjectDraft(
    val objectId: String,
    val tag: String,
    val reference: ObjectReference,
)

interface ObjectStore {
    suspend fun create(draft: ObjectDraft): TraceResult<ObjectReference>
    suspend fun get(objectId: String): TraceResult<ObjectReference>
    suspend fun getAllReferences(): TraceResult<List<ObjectReference>>
    suspend fun delete(objectId: String): TraceResult<Unit>
}

interface SightingStore {
    suspend fun insert(sighting: Sighting): TraceResult<Sighting>
    suspend fun update(sighting: Sighting): TraceResult<Sighting>
    suspend fun getLatest(objectId: String): TraceResult<Sighting?>
    suspend fun getTimeline(objectId: String, limit: Int = 50): TraceResult<List<Sighting>>
    suspend fun getPending(limit: Int = 100): TraceResult<List<Sighting>>
}

enum class SecureAssetType { REFERENCE_IMAGE, SIGHTING_EVIDENCE }

data class SecureAsset(
    val assetId: String,
    val type: SecureAssetType,
    val mimeType: String,
    val createdAtEpochMillis: Long,
)

interface SecureAssetStore {
    suspend fun write(
        type: SecureAssetType,
        plaintext: ByteArray,
        mimeType: String,
    ): TraceResult<SecureAsset>

    suspend fun read(assetId: String): TraceResult<ByteArray>
    suspend fun delete(assetId: String): TraceResult<Unit>
}
